docs: [FC-0074] add comparison table between the framework and forking openedx (#617)
openedx/docs.openedx.org: source/developers/concepts/hooks_extension_framework.rst

|                                                                    | **Open edX Events and Filters**                                                                               | **Forking**                                                                                                                        |
+====================================================================+===============================================================================================================+====================================================================================================================================+
| **Purpose**                                                        | Allow extending functionality of the Open edX platform without modifying the                                  | Modify the code directly to meet specific needs.                                                                                   |

> |                                                                    | core codebase but by leveraging and configurations.                                                           |                                                                                                                                    |

Author: I just realized I missed the word plugins in _by leveraging plugins and configurations_

| **Disadvantages**                                                  | - Can increase complexity in debugging and maintenance because it adds more processing layers.                | - Separation from the core makes long-term maintenance challenging.                                                                |
|                                                                    | - Limited by existing events and filters available in the Open edX ecosystem.                                 | - High maintenance burden for upgrades and compatibility.                                                                          |
|                                                                    | - Might increase complexity in plugins for highly customized features.                                        | - Requires deep understanding of the platform to modify it correctly.                                                              |

> |                                                                    | - Might introduce an edx-platform dependency into plugins, mainly when using filters.                         |                                                                                                                                    |

Author: When implementing new features in plugins, even though you could be using an extension point which should provide enough data about what's being extended, e.g., the enrollment filter provides the `user`, `course_key`, and `enrollment mode`, we might want to perform operations on the data like getting the `course overview` object that inevitably could result in importing edx-platform code into plugins. This is an antipattern, and it's strongly not recommended, but there's no current alternative for doing that kind of operation (this might be solved by learning core in the long term). Is this too specific to add here?

> |                                                                    | - Might introduce an edx-platform dependency into plugins, mainly when using filters.                         |                                                                                                                                    |

Reviewer: How does this sound as a summary / a little bit more detail & clarity?

```suggestion
|                                                                    | - Although it is not recommended to do so, a plugin might introduce an edx-platform dependency, mainly when using filters, when there is no alternative for getting certain types of data.                         |                                                                                                                                    |
```

> |                                                                    | - Might introduce an edx-platform dependency into plugins, mainly when using filters.                         |                                                                                                                                    |

Author: Yes! Thank you. I implemented the suggestion here: https://github.com/openedx/docs.openedx.org/pull/617/commits/d4266da384e73d2dd6713fe1a897503a19c0ab4c

+--------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------+

> | **Performance Impact**                                             | Medium to high, as events and filters execute within the caller process. Although events are designed so      | High, as changes directly impact the Open edX platform (edx-platform or other IDAs) and may require additional testing, monitoring,|

Author: I think both the framework and forking could behave similarly under this category cause this mainly depends on how performant the solution is. I will remove this category and move the _it may require additional testing..._ from the fork column to the implementation effort row.

|                                                                    | receivers can trigger asynchronous tasks for improved performance.                                            | and performance considerations of new added code and for the integration itself.                                                   |
+--------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------+

> | **Community Compatibility**                                        | High, as the community encourage reusable implementations and the contribution of new extension points.       | Low, as changes in forks are usually highly customized not designed for others to reuse.                                           |

Reviewer: ```suggestion
| **Community Compatibility**                                        | High, as the community encourages reusable implementations and the contribution of new extension points.       | Low, as changes in forks are usually highly customized and not designed for others to reuse.                                           |
```

+--------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------+

> | **Implementation Effort**                                          | Low to medium, depending on the complexity of the receiver logic and filter pipelines.                        | High, as modifications require expertise in the platform and increase technical debt.                                              |

Reviewer: ```suggestion
| **Implementation Effort**                                          | Low to medium, depending on the complexity of the receiver logic and filter pipelines.                        | High, as modifications require expertise in the platform and increases technical debt.                                              |
```

We encourage you to review the list of use cases for events and filters to draw inspiration from real-life scenarios and see if your use case aligns with any of them. Also, maybe your feature can be implementing using the framework but there's not an available event or filter for it yet. In that case, consider proposing a new event or filter to the community!

The Framework vs Forking the Open edX Platform
**********************************************

> The Hooks Extension Framework is designed to provide a stable and maintainable way to extend the Open edX platform without modifying its core. This approach is in contrast to forking the platform, which involves creating a separate version that diverges from the core codebase. Here is a comparison table of the two approaches:

Reviewer 2: This approach is in contrast to forking -> sounds a bit weird to me.

How about: This approach stands in opposition to the practice of forking ... ?

|                                                                    | **Open edX Events and Filters**                                                                               | **Forking**                                                                                                                        |
+====================================================================+===============================================================================================================+====================================================================================================================================+
| **Purpose**                                                        | Allow extending functionality of the Open edX platform without modifying the                                  | Modify the code directly to meet specific needs.                                                                                   |

> |                                                                    | core codebase but by leveraging plugins and configurations.                                                   |                                                                                                                                    |

Reviewer 2: ```suggestion
|                                                                    | core codebase by leveraging plugins and configurations.                                                   |                                                                                                                                    |
```

Author: Thanks. I completely missed that.

+--------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------+
| **Advantages**                                                     | - More decoupled implementations.                                                                             | - Full control of the platform behavior.                                                                                           |
|                                                                    | - Decreased maintenance burden of the platform, maintenance efforts are centered on plugins.                  | - No restrictions from what can be changed.                                                                                        |

> |                                                                    | - Enables reusable and configurable behavior across the platform.                                             | - Suitable for highly customized use cases.                                                                                        |

Reviewer 2: ```suggestion
|                                                                    | - Enables reusable and configurable behavior across the platform.                                             | - Suitable for extremely customized use cases.                                                                                        |
```
I think that hooks framework already allows for high customization. IMHO the modification must be extreme to warrant forking where hooks can do it. The more regular reason is when the platform does not support hooks in the place you want to customize.

Reviewer 2: Another point in this line would be:

> Enables reusable and configurable behavior across the platform

I think is worth splitting in two:

- Enables reusable behavior across the ecosystem and community.
- Enables configurable behavior across the platform

Author: > I think that hooks framework already allows for high customization. IMHO the modification must be extreme to warrant forking where hooks can do it. The more regular reason is when the platform does not support hooks in the place you want to customize.

I agree. I added a note suggesting to look into how the framework can help before forking: https://github.com/openedx/docs.openedx.org/pull/617/commits/67bd6f7c4060421d7b2dea42c6affffa8a8d6acd#diff-01d0bcd5d9113aaa8a7f3fde94c7e5ca100265eebebbad867dbf4de21123d451R193

I also split the sentence you mentioned into two for more clarity: https://github.com/openedx/docs.openedx.org/pull/617/commits/67bd6f7c4060421d7b2dea42c6affffa8a8d6acd#diff-01d0bcd5d9113aaa8a7f3fde94c7e5ca100265eebebbad867dbf4de21123d451R172-R173

|                                                                    | - Easier to maintain long-term due to backward compatibility and versioning policies.                         |                                                                                                                                    |
+--------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------+

> | **Disadvantages**                                                  | - Can increase complexity in debugging and maintenance because it adds more processing layers.                | - Separation from the core makes long-term maintenance challenging.                                                                |

Reviewer 2: I think we should call it "separation from the upstream project" because core has a different meaning here.

Author: That's more clear. Thank you.

https://github.com/openedx/docs.openedx.org/pull/617/commits/67bd6f7c4060421d7b2dea42c6affffa8a8d6acd#diff-01d0bcd5d9113aaa8a7f3fde94c7e5ca100265eebebbad867dbf4de21123d451R177

|                                                                    | - Limited by existing events and filters available in the Open edX ecosystem, although new events and         | - High maintenance burden for upgrades and compatibility.                                                                          |
|                                                                    |   filters can be proposed and added to the framework.                                                         | - Requires manual integration with the Open edX platform for each new release.                                                     |

> |                                                                    | - Might increase complexity in plugins for highly customized features.                                        | - Requires deep understanding of the platform to modify it correctly.                                                              |

Reviewer 2: > Requires deep understanding of the platform to modify it correctly

I would argue that modifying correctly means using the open-closed principle which is not being used if the road was a fork. 

I would list however the "not being able to reuse by the community" as part of the disadvantages.

Reviewer 2: Aso a disadvantage of fork:

- difficult to test in isolation

Author: > I would argue that modifying correctly means using the open-closed principle which is not being used if the road was a fork.

I removed the "correctly" from the sentence. I do think that to modify the platform, you need a deep understanding of the components you'd be modifying, mainly because of the implications of the changes. But one can argue that you'd need the same to add new extension points, so I'll drop the line altogether. 

> I would list however the "not being able to reuse by the community" as part of the disadvantages.

Added! Thanks https://github.com/openedx/docs.openedx.org/pull/617/commits/67bd6f7c4060421d7b2dea42c6affffa8a8d6acd#diff-01d0bcd5d9113aaa8a7f3fde94c7e5ca100265eebebbad867dbf4de21123d451R181

|                                                                    | - Although it is not recommended to do so, a plugin developer might introduce an edx-platform dependency,     |                                                                                                                                    |
|                                                                    |   mainly when using filters, when there is no alternative for getting certain types of data.                  |                                                                                                                                    |
+--------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------+------------------------------------------------------------------------------------------------------------------------------------+

> | **Community Compatibility**                                        | High, as the community encourages reusable implementations and the contribution of new extension points.      | Low, as changes in forks are usually highly customized not designed for others to reuse.                                           |

Reviewer 2: > Low, as changes in forks are usually highly customized not designed for others to reuse.

And discoverability is low. Most probably no one will ever know "you did that".

Author: I added a note about discoverability here: https://github.com/openedx/docs.openedx.org/pull/617/commits/67bd6f7c4060421d7b2dea42c6affffa8a8d6acd#diff-01d0bcd5d9113aaa8a7f3fde94c7e5ca100265eebebbad867dbf4de21123d451R185
